### PR TITLE
JIT: Add support for frozen structs in Swift reverse pinvokes

### PR DESCRIPTION
**Pull request by jakobbotsch**
This adds the final support for frozen structs in UCO methods.

This passes 2500 auto generated tests locally on both macOS x64 and arm64. This PR includes only 100 tests.

Frozen struct parameters are handled via the new ABI representation added in #100138. When such a parameter exists we always allocate space for it on the local stack frame. The struct is then reassembled from its passed constituents as the first thing in the codegen.

One complication is that there can be an arbitrary amount of codegen to handle this reassembling. We cannot handle an arbitrary amount of codegen in the prolog, so the reassembling is handled in two places. First, since the amount of register passed data is limited, we handle those in the prolog (which frees them up to be used for other things). If some pieces were passed on the stack the JIT then ensures that there is a scratch BB and generates the code to reassemble the remaining parts as the first thing in the scratch BB.

Since Swift structs can be passed by reference in certain cases this PR also enables `FEATURE_IMPLICIT_BYREFS` for SysV x64 to handle those cases. Depending on the TP impact we can refine some of the ifdefs around this.

TODO: 
- [x] Fix lvRefCnt() == 0 case

### CODE REVIEW

**Comment by jakobbotsch** [comment]
I've optimized the throughput a bit, but there are still some large ish TP regressions on SysV x64 (which probably are more impactful in MinOpts). They come from enabling implicit byrefs on that target. Locally the current breakdown looks like
```scala
Base: 239570519361, Diff: 239967442865, +0.1657%

112977529 : +10.45%  : 13.30% : +0.0472% : public: enum Compiler::fgWalkResult __cdecl LocalAddressVisitor::PreOrderVisit(struct GenTree **, struct GenTree *)                                                                                                                                       
87504609  : +44.17%  : 10.30% : +0.0365% : private: struct GenTree * __cdecl Compiler::fgMorphExpandLocal(struct GenTreeLclVarCommon *)                                                                                                                                                              
56537866  : +3.48%   : 6.65%  : +0.0236% : public: void __cdecl Compiler::lvaComputeRefCounts(bool, bool)                                                                                                                                                                                            
29348940  : +4.41%   : 3.45%  : +0.0123% : public: void __cdecl LclVarDsc::incRefCnts(double, class Compiler *, enum RefCountState, bool)                                                                                                                                                            
26360300  : +16.28%  : 3.10%  : +0.0110% : public: bool __cdecl Compiler::fgAddrCouldBeNull(struct GenTree *)                                                                                                                                                                                        
22399172  : NA       : 2.64%  : +0.0093% : VisitEHSuccs<0,`LinearScan::buildIntervals<1>'::`117'::<lambda_1> >                                                                                                                                                                                       
19685105  : +1.85%   : 2.32%  : +0.0082% : public: void __cdecl LinearScan::buildIntervals<1>(void)                                                                                                                                                                                                  
18636361  : NA       : 2.19%  : +0.0078% : public: void __cdecl Compiler::lvaClassifyParameterABI(void)                                                                                                                                                                                              
15770012  : +5.12%   : 1.86%  : +0.0066% : private: struct GenTree * __cdecl Compiler::fgMorphLeaf(struct GenTree *)                                                                                                                                                                                 
14498187  : +8.51%   : 1.71%  : +0.0061% : protected: void __cdecl CodeGen::genFnPrologCalleeRegArgs(enum _regNumber_enum, bool *, struct RegState *)                                                                                                                                                
10892949  : +0.59%   : 1.28%  : +0.0045% : protected: void __cdecl CodeGen::genCodeForBBlist(void)                                                                                                                                                                                                   
10532735  : NA       : 1.24%  : +0.0044% : private: enum PhaseStatus __cdecl Compiler::fgRetypeImplicitByRefArgs(void)                                                                                                                                                                               
10054272  : +5.68%   : 1.18%  : +0.0042% : public: unsigned int __cdecl emitter::emitInsSizeSVCalcDisp(struct emitter::instrDesc *, unsigned __int64, int, int)                                                                                                                                      
5734540   : NA       : 0.67%  : +0.0024% : public: bool __cdecl Compiler::lvaParamShouldHaveLocalStackSpace(unsigned int)                                                                                                                                                                            
5625675   : +0.19%   : 0.66%  : +0.0023% : private: struct GenTree * __cdecl Compiler::fgMorphSmpOp(struct GenTree *, struct Compiler::MorphAddrContext *, bool *)                                                                                                                                   
4311764   : +22.53%  : 0.51%  : +0.0018% : `Compiler::compCompile'::`2'::<lambda_3>::operator()                                                                                                                                                                                                      
4062020   : +1.79%   : 0.48%  : +0.0017% : private: bool __cdecl Compiler::fgForwardSubStatement(struct Statement *)                                                                                                                                                                                 
3785434   : +0.64%   : 0.45%  : +0.0016% : private: void __cdecl TreeLifeUpdater<1>::UpdateLifeVar(struct GenTree *, struct GenTreeLclVarCommon *)                                                                                                                                                   
2036910   : +2.30%   : 0.24%  : +0.0009% : private: void __cdecl Compiler::fgKillDependentAssertions(unsigned int)                                                                                                                                                                                   
1973170   : +0.63%   : 0.23%  : +0.0008% : public: struct CORINFO_CLASS_STRUCT_* __cdecl Compiler::gtGetClassHandle(struct GenTree *, bool *, bool *)                                                                                                                                                
1563661   : +3.48%   : 0.18%  : +0.0007% : public: void __cdecl Compiler::lvaSetStruct(unsigned int, class ClassLayout *, bool)                                                                                                                                                                      
1561885   : +1.35%   : 0.18%  : +0.0007% : public: void __cdecl LinearScan::buildIntervals<0>(void)                                                                                                                                                                                                  
1261162   : +4.19%   : 0.15%  : +0.0005% : private: struct GenTree * __cdecl Compiler::fgMorphIndexAddr(struct GenTreeIndexAddr *)                                                                                                                                                                   
1258128   : +1.72%   : 0.15%  : +0.0005% : public: struct CodeGenInterface::siVarLoc __cdecl CodeGenInterface::getSiVarLoc(class LclVarDsc const *, unsigned int) const                                                                                                                              
1100668   : +25.12%  : 0.13%  : +0.0005% : public: enum var_types __cdecl Compiler::getArgTypeForStruct(struct CORINFO_CLASS_STRUCT_*, enum Compiler::structPassingKind *, bool, unsigned int)                                                                                                       
1001286   : +5.95%   : 0.12%  : +0.0004% : private: struct GenTree * __cdecl Compiler::fgMorphPotentialTailCall(struct GenTreeCall *)                                                                                                                                                                
908520    : NA       : 0.11%  : +0.0004% : public: bool __cdecl Compiler::lvaHasAnySwiftStackParamToReassemble(void)                                                                                                                                                                                 
898999    : +1.51%   : 0.11%  : +0.0004% : public: virtual void __cdecl CSE_HeuristicCommon::PerformCSE(class CSE_Candidate *)                                                                                                                                                                       
849770    : +4.62%   : 0.10%  : +0.0004% : public: void __cdecl Compiler::lvaUpdateArgsWithInitialReg(void)                                                                                                                                                                                          
-9772408  : -0.69%   : 1.15%  : -0.0041% : private: int __cdecl LinearScan::BuildNode(struct GenTree *)                                                                                                                                                                                              
-11475144 : -0.64%   : 1.35%  : -0.0048% : memset                                                                                                                                                                                                                                                    
-15704816 : -63.01%  : 1.85%  : -0.0066% : public: void __cdecl Compiler::lvaInitArgs(struct InitVarDscInfo *)                                                                                                                                                                                       
-21137984 : -1.21%   : 2.49%  : -0.0088% : private: class RefPosition * __cdecl LinearScan::newRefPosition(class Interval *, unsigned int, enum RefType, struct GenTree *, unsigned __int64, unsigned int)                                                                                           
-22399172 : -100.00% : 2.64%  : -0.0093% : VisitEHSuccs<0,`LinearScan::buildIntervals<1>'::`104'::<lambda_1> >                                                                                                                                                                                                                                                                                                                                                                             
```

I don't see a good way to avoid this impact, so I think we'll just have to live with it.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @amanasifkhalid 

I think this will need a few fixups after #100429 is merged. 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=629969&view=ms.vss-build-web.run-extensions-tab). Enabling implicit byrefs on SysV x64 results in some large TP regressions with an MSVC compiled cross compiler, but the clang compiled one doesn't look bad. Also some TP improvements on other platforms from some of the optimization work I did around the implicit byrefs as a result.

**Review by amanasifkhalid — APPROVED** [review]
LGTM assuming runtime/stress pipelines pass. Thanks!

File: src/coreclr/jit/codegencommon.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp:6350]: Wouldn't it be a bug for this to be an `else if`? Is anything stopping us from having both`SwiftSelf` and `SwiftError`?

Comment (amanasifkhalid) [reply src/coreclr/jit/codegencommon.cpp:6350]: Ah you're right, for some reason I misread this as comparing `lvaSwiftErrorArg` to a local variable number.

File: src/coreclr/jit/abi.cpp
Comment (jakobbotsch) [reply src/coreclr/jit/abi.cpp:113]: We generally normalize up to at least 4 bytes on all loads, and zero extensions sometimes have smaller encodings than sign extensions on xarch (similar change was made in #86631).

**Comment by jakobbotsch** [comment]
I'll kick off the stress pipelines again just to be safe after the merge of #100429.

**Comment by jakobbotsch** [comment]
Also cc @jkoritzinsky, @jkurdek, @matouskozak, @kotlarmilos. With this PR I think we'll be feature complete on the Swift support in RyuJIT (until #100543 makes it through review, although I expect it will be straightforward to add support for those).

**Comment by jakobbotsch** [comment]
Failures are #100476 and #99810

**Pull request merged**
